SPEC-annex-A-filesystem-storage.md

```diff
@@ -0,0 +1,49 @@
+## Annex A
+
+Annex A documents known methods of persisting GitBOM Documents to various stores.
+### GitBOM Document persistence by a Build Tool to its local filesystem
+
+If a build tool persists GitBOM information to its local filesystem, the build tool should write out the GitBOM Document to ```${GITBOM_DIR}/objects/${Artifact Identifier Type uri prefix with ':' replaced by '/'}/${GitBOMID:0:2}/${GitBOMID:2:}``` where ```${GitBOMID}``` is the GitBOM id in lowercase hexidecimal without leading zeros suppressed.
+
+Example:
+
+```
+.adg/objects/gitoid/blob/sha1/0e/8efd4cdf0d5bafcfcae658c2662a73b199b301
+```
+
+#### Build tool persistence of artifact to GitBOM Document mapping
+
+A build tool should choose to persist a mapping between artifacts and their corresponding GitBOM documents.  If it chooses
+to do so it should for each artifact persist a symlink:
+
+```${GITBOM_DIR}/a2g/${Artifact Identifier Type uri prefix with ':' replaced by '/'}/${artifact id}:0:2}/${artifact id:2:} -> ${GITBOM_DIR}/objects/${Artifact Identifier Type uri prefix with ':' replaced by '/'}/${GitBOMID:0:2}/${GitBOMID:2:}```
```

Comment:
Need some number of `../../` here to actually point to the GitBOM documents.

Reply:
I like this symlink farm idea. :-)

I have implemented this symlink farm in gcc to persist the artifact-id to bom_id mapping:
``` ls -tl *
symlinks/9d9db0273466bda3693c2e5515f84b190ac249f413c81e6380609bc202136910.a2g.symlink -> ../adg/ae2e0319378aacd8338ac03ce2adfa3f04f159bca9fd3afb771b0dde2c0275ed
adg/ae2e0319378aacd8338ac03ce2adfa3f04f159bca9fd3afb771b0dde2c0275ed
```
This does work: without embedding gitoid or bom_id into generated ELF files, this symlink helps find the gitBOM doc of an artifact easily.

However, there is one race condition concern for parallel builds on distributed systems:

Generating gitBOM doc involves a few lookups of some symlinks created by previous build steps. For example, when linking hello executable, the symlinks of the hello1.o, hello2.o, ..., and helloN.o object files are queried. This may not work well with parallel builds. What if the hello1.o file is generated but the symlink for hello1.o is not yet created when invoking the linking step? Then the bom_id lookup for hello1.o will return empty, thus generating incorrect gitBOM doc. So there seems some probability for this race condition.

Note it seems hard to add an extra Makefile dependency of hello1.o.a2g.symlink to avoid this race condition, since this symlink may actually not exist, which is valid scenario.

In contrast, the record-hash-only mode does not have this issue, because at each build step, we only record the hashes of output and input files, which do not rely on any previous build steps. Thus there is no race condition.

Let me know what you think of this race condition issue, and what suggestions.

Reply:
I have implemented this symlink farm in gcc, and done some experiments with parallel builds. It seems fine in my implementation: The same process creates the output file and its symlink to ADG. As long as this process does not exit, the new process which depends on this process's output file seems to not start. Therefore, so far so good with my gcc. Anyway, I will keep an eye on this concern in my future experiments.

Reply:
Your findings are consistent with my expectations.  As long as writing the artifact and its GitBOM document are 'atomic' at the level of 'event consumed by downstream consumer' we should be fine.  So as long as the triggering event for the consumer is something like 'build tool exits' the 'atomicity' is preserved and we should be good on race conditions.